File: test/rpc1_test.go
package test

import (
	"context"
	"errors"
	"fmt"
	"io"
	"testing"
	"time"

	"github.com/stretchr/testify/require"
	"ne43.pub/go-snowpack-rpc/rpc"
)

type ErrString string
type ErrInt int64

func (e ErrString) Error() string {
	return fmt.Sprintf("string error: %s", string(e))
}

func (e ErrString) toString() string { return string(e) }

func (i ErrInt) Error() string {
	return fmt.Sprintf("int error: %d", int64(i))
}

func (i ErrInt) toInt() int64 { return int64(i) }

func errorToStatus(e error) Status {
	if e == nil {
		return NewStatusWithOk()
	}
	switch te := e.(type) {
	case ErrString:
		return NewStatusWithEstring(te.toString())
	case ErrInt:
		return NewStatusWithEint(te.toInt())
	case rpc.ProtocolV2NotFoundError:
		return NewStatusWithEprotonotfound(uint64(te.U))
	case rpc.MethodV2NotFoundError:
		return NewStatusWithEmethodnotfound(MethodV2{
			Proto:  uint64(te.ProtID),
			Method: uint64(te.Method),
			Name:   te.ProtName,
		})
	default:
		return NewStatusWithEgeneric(te.Error())
	}
}

func errorUnwrapper(status Status) error {
	sc, err := status.GetSc()
	if err != nil {
		return err
	}
	switch sc {
	case StatusCode_OK:
		return nil
	case StatusCode_EINT:
		return ErrInt(status.Eint())
	case StatusCode_ESTRING:
		return ErrString(status.Estring())
	case StatusCode_EGENERIC:
		return errors.New(status.Egeneric())
	case StatusCode_EPROTONOTFOUND:
		return rpc.NewProtocolV2NotFoundError(rpc.ProtocolUniqueID(status.Eprotonotfound()))
	case StatusCode_EMETHODNOTFOUND:
		m := status.Emethodnotfound()
		return rpc.NewMethodV2NotFoundError(rpc.ProtocolUniqueID(m.Proto), rpc.Position(m.Method), m.Name)
	default:
		return fmt.Errorf("generic error %d", int(sc))
	}
}

type PingServer struct {
}

func (p PingServer) PingUint(ctx context.Context, u uint64) (ret uint64, err error) {
	return u + 10, nil
}

func (p PingServer) PingText(ctx context.Context, s string) (ret string, err error) {
	return s + "_yo", nil
}

func (p PingServer) PingSmoosh(ctx context.Context, arg PingSmooshArg) (ret Smoosh, err error) {
	return Smoosh{
		I: arg.I + 3,
		T: fmt.Sprintf("%s%d", arg.T, arg.I),
	}, nil
}

func (p PingServer) PingPuke(ctx context.Context, arg PingPukeArg) (err error) {
	if arg.I != nil {
		return ErrInt(*arg.I)
	}
	if arg.T != nil {
		return ErrString(*arg.T)
	}
	return errors.New("generic error")
}

func (p PingServer) BlobTorture(ctx context.Context, arg [][]byte) ([][]byte, error) {
	for i, j := 0, len(arg)-1; i < j; i, j = i+1, j-1 {
		arg[i], arg[j] = arg[j], arg[i]
	}
	return arg, nil
}

func (p PingServer) PingSmooshes(ctx context.Context, arg []Smoosh) ([]Smoosh, error) {
	for i, j := 0, len(arg)-1; i < j; i, j = i+1, j-1 {
		arg[i], arg[j] = arg[j], arg[i]
	}
	return arg, nil
}

func (p PingServer) GetNow(ctx context.Context) (ret UnixTime, err error) {
	return UnixTime(time.Now().UnixMilli()), nil
}

func (p PingServer) ArgRename(ctx context.Context, a MyArg) (string, error) {
	return "", nil
}

func (p PingServer) ErrorWrapper() func(error) Status {
	return errorToStatus
}

func (p PingServer) CheckArgHeader(ctx context.Context, h Header) error {
	if h.Vers != 2 {
		return errors.New("bad version; can only handle v2")
	}
	return nil
}

func (p PingServer) MakeResHeader() Header {
	return Header{
		Vers: 3,
	}
}

var _ PingInterface = &PingServer{}

func TestPing1(t *testing.T) {

	// Turn debugging on for now
	dbgHook := func(s string, args ...interface{}) {}
	lf := rpc.NewSimpleLogFactory(rpc.NilLogOutput{}, nil)

	loopack := rpc.NewLoopbackListener(dbgHook)

	// Fork a listener in the background
	serverExitCh := make(chan struct{})

	maxSz := int32(0x10000)

	wrapErrorFunc := PingMakeGenericErrorWrapper(errorToStatus)

	go (func() {
		conn, err := loopack.Accept()
		require.NoError(t, err)
		xp := rpc.NewTransport(context.Background(), conn, lf, nil, wrapErrorFunc, maxSz)
		srv := rpc.NewServer(xp, wrapErrorFunc)
		var ps PingServer
		srv.RegisterV2(PingProtocol(ps))
		ch := srv.Run()
		<-ch
		err = srv.Err()
		require.Error(t, err)
		require.Equal(t, io.EOF, err)
		close(serverExitCh)
	})()

	ctx := context.Background()
	conn, err := loopack.Dial(ctx)
	require.NoError(t, err)

	xp := rpc.NewTransport(context.Background(), conn, lf, nil, wrapErrorFunc, maxSz)
	gcli := rpc.NewClient(xp, nil, nil)
	cli := PingClient{
		Cli:            gcli,
		ErrorUnwrapper: errorUnwrapper,
		MakeArgHeader: func() Header {
			return Header{
				Vers: 2,
			}
		},
		CheckResHeader: func(ctx context.Context, h Header) error {
			if h.Vers != 3 {
				return errors.New("bad version; can only handle v3")
			}
			return nil
		},
	}

	ret, err := cli.PingUint(context.Background(), 5)
	require.NoError(t, err)
	require.Equal(t, uint64(15), ret)

	s, err := cli.PingText(context.Background(), "blipper")
	require.NoError(t, err)
	require.Equal(t, "blipper_yo", s)

	sm, err := cli.PingSmoosh(context.Background(), PingSmooshArg{I: -10, T: "lucky"})
	require.NoError(t, err)
	require.Equal(t, Smoosh{I: -7, T: "lucky-10"}, sm)

	i := int64(10)
	err = cli.PingPuke(context.Background(), PingPukeArg{I: &i})
	require.Equal(t, ErrInt(i), err)

	s = "bingo"
	err = cli.PingPuke(context.Background(), PingPukeArg{T: &s})
	require.Equal(t, ErrString(s), err)

	err = cli.PingPuke(context.Background(), PingPukeArg{})
	require.Equal(t, errors.New("generic error"), err)

	res, err := cli.BlobTorture(context.Background(), [][]byte{
		[]byte("all"),
		[]byte("dogs"),
		[]byte("have"),
		[]byte("fleas"),
	})
	require.NoError(t, err)
	require.Equal(t, []byte("have"), res[1])
	require.Equal(t, []byte("all"), res[3])

	smooshesRes, err := cli.PingSmooshes(context.Background(), []Smoosh{
		{I: 1, T: "one"},
		{I: 2, T: "two"},
		{I: 3, T: "three"},
	})
	require.NoError(t, err)
	require.Equal(t, []Smoosh{
		{I: 3, T: "three"},
		{I: 2, T: "two"},
		{I: 1, T: "one"},
	}, smooshesRes)

	poogCli := PoogClient{Cli: gcli, ErrorUnwrapper: errorUnwrapper}
	err = poogCli.PoogUint(context.Background())
	require.Equal(t, rpc.NewProtocolV2NotFoundError(PoogProtocolID), err)

	pingV2Cli := PingV2Client{Cli: gcli, ErrorUnwrapper: errorUnwrapper}
	err = pingV2Cli.PingPike(context.Background())
	require.Equal(t, rpc.NewMethodV2NotFoundError(PingProtocolID, 50, "ping"), err)

	cli.MakeArgHeader = func() Header {
		return Header{
			Vers: 1,
		}
	}
	_, err = cli.PingUint(context.Background(), 5)
	require.Error(t, err)
	require.Equal(t, errors.New("bad version; can only handle v2"), err)

	cli.CheckResHeader = func(ctx context.Context, h Header) error {
		if h.Vers != 2 {
			return errors.New("bad version; can only handle v2")
		}
		return nil
	}
	_, err = cli.PingUint(context.Background(), 5)
	require.Error(t, err)
	require.Equal(t, errors.New("bad version; can only handle v2"), err)

	conn.Close()
	<-serverExitCh

}

File: test/rpc1.test.ts
import { UnixTime } from './common'
import {
    MethodV2,
    MyArg,
    pingClient,
    pingV2Client,
    pingV2ProtocolID,
    poogProtocolID,
    pingInterface,
    pingProtocol,
    pingPukeArg,
    pingSmooshArg,
    Smoosh,
    Status,
    StatusCode,
    poogClient,
} from './rpc1'

import {
    NetConnOpts,
    Client,
    Handler,
    Server,
    ErrMethodNotFound,
    ErrProtocolNotFound,
    newProtocolUniqueID,
} from 'snowpack'
const suppressLogging = true

class ErrInt extends Error {
    i: bigint
    constructor(i: bigint) {
        super(`ErrInt: ${i}`)
        this.i = i
    }
}

class ErrString extends Error {
    s: string
    constructor(s: string) {
        super(`ErrString: ${s}`)
        this.s = s
    }
}

function errorToStatus(e: Error | null): Status {
    if (e === null) {
        return Status.newWithOK()
    }
    if (e instanceof ErrInt) {
        return Status.newWithEINT(e.i)
    } else if (e instanceof ErrString) {
        return Status.newWithESTRING(e.s)
    } else if (e instanceof ErrMethodNotFound) {
        const ret = Status.newWithEMETHODNOTFOUND(
            new MethodV2(e.protocol.value, BigInt(e.method), 'n/a')
        )
        return ret
    } else if (e instanceof ErrProtocolNotFound) {
        const ret = Status.newWithEPROTONOTFOUND(e.protocol.value)
        return ret
    } else {
        return Status.newWithEGENERIC(e.message)
    }
}

function statusToError(s: Status): Error | null {
    const tag = s.sc
    switch (s.sc.value) {
        case StatusCode.values.OK:
            return null
        case StatusCode.values.EINT:
            return new ErrInt(s.EINT)
        case StatusCode.values.ESTRING:
            return new ErrString(s.ESTRING)
        case StatusCode.values.EPROTONOTFOUND:
            return new ErrProtocolNotFound(
                newProtocolUniqueID(s.EPROTONOTFOUND)
            )
        case StatusCode.values.EMETHODNOTFOUND:
            return new ErrMethodNotFound(
                newProtocolUniqueID(s.EMETHODNOTFOUND.proto),
                Number(s.EMETHODNOTFOUND.method)
            )
        default:
            return new Error(s.EGENERIC)
    }
}

class PingServer extends Handler implements pingInterface {
    async pingUint(arg: bigint): Promise<bigint> {
        return arg + 10n
    }
    async pingText(arg: string): Promise<string> {
        return arg + '_yo'
    }
    async pingSmoosh(arg: pingSmooshArg): Promise<Smoosh> {
        return new Smoosh(arg.i + 3n, `${arg.t}${arg.i}`)
    }
    async pingPuke(arg: pingPukeArg): Promise<void> {
        if (arg.i !== undefined && arg.i !== null) {
            throw new ErrInt(arg.i)
        }
        if (arg.t !== undefined && arg.t !== null) {
            throw new ErrString(arg.t)
        }
        throw new Error('generic')
    }
    async blobTorture(arg: Uint8Array[]): Promise<Uint8Array[]> {
        return arg.reverse()
    }
    async pingSmooshes(arg: Smoosh[]): Promise<Smoosh[]> {
        return arg.reverse()
    }
    async argRename(arg: MyArg): Promise<string> {
        return arg.a.toString(10)
    }
    async getNow(): Promise<UnixTime> {
        return UnixTime.import(Date.now())
    }
    errorWrapper(arg: Error | null): Status {
        return errorToStatus(arg)
    }
}
var port: number | undefined
var srv: Server<PingServer> | undefined

async function gConnect(): Promise<Client> {
    if (!port) {
        throw new Error('expected port to be non-null')
    }
    const nco = new NetConnOpts('localhost', port, {
        nullLog: suppressLogging,
    })
    const gcli = new Client(nco)
    await gcli.connect()
    return gcli
}

async function connect(): Promise<[pingClient, Client]> {
    const gcli = await gConnect()
    const cli = new pingClient(gcli, statusToError)
    return [cli, gcli]
}

describe('rpc1', () => {
    beforeAll(async () => {
        srv = new Server(
            {
                port: 0,
                host: 'localhost',
                path: undefined,
                nullLog: suppressLogging,
            },
            [pingProtocol],
            PingServer
        )
        await srv.listen()
        port = srv.port
    })

    it('should handle simple successful RPC calls', async () => {
        const [cli, gcli] = await connect()

        const res = await cli.pingUint(11n)
        expect(res).toBe(21n)

        const res2 = await cli.pingText('blippy')
        expect(res2).toBe('blippy_yo')

        const res3 = await cli.pingSmoosh(new pingSmooshArg(11n, 'blippy'))
        expect(res3).toEqual(new Smoosh(14n, 'blippy11'))

        const arg4 = 0x1234567890abcdefn
        const res4 = await cli.pingUint(arg4)
        expect(res4).toBe(arg4 + 10n)

        const arg5 = ['all', 'dogs', 'have', 'fleas'].map((s) => Buffer.from(s))
        const res5 = await cli.blobTorture(arg5)
        expect(res5).toEqual(arg5.reverse())

        gcli.close()
    })
    it('should handle RPCs with arrays of objects', async () => {
        const [cli, gcli] = await connect()

        const arg = [
            new Smoosh(1n, 'a'),
            new Smoosh(2n, 'b'),
            new Smoosh(3n, 'c'),
        ]
        const res = await cli.pingSmooshes(arg)
        expect(res).toEqual(arg.reverse())

        gcli.close()
    })
    it('should tell time', async () => {
        const [cli, gcli] = await connect()

        const now = Date.now()
        const res = await cli.getNow()
        const then = Number(res.export())
        expect(Math.abs(now - then)).toBeLessThan(10000)

        gcli.close()
    })

    it('should handle simple unsuccessful RPC calls with error import/export', async () => {
        const [cli, gcli] = await connect()

        try {
            await cli.pingPuke(new pingPukeArg(11n, null))
            throw new Error('expected pingPuke to throw')
        } catch (e) {
            expect(e).toBeInstanceOf(ErrInt)
            const ei = e as ErrInt
            expect(ei.i).toBe(11n)
        }

        try {
            await cli.pingPuke(new pingPukeArg(null, 'bluey'))
            throw new Error('expected pingPuke to throw')
        } catch (e) {
            expect(e).toBeInstanceOf(ErrString)
            const es = e as ErrString
            expect(es.s).toBe('bluey')
        }
        gcli.close()
    })

    it('should handle method not found', async () => {
        const gcli = await gConnect()
        const cli = new pingV2Client(gcli, statusToError)
        try {
            await cli.pingPike()
            throw new Error('expected pingPike to throw')
        } catch (e) {
            expect(e).toBeInstanceOf(ErrMethodNotFound)
            const es = e as ErrMethodNotFound
            expect(es.protocol.value).toBe(pingV2ProtocolID.value)
            expect(es.method).toBe(50)
        } finally {
            gcli.close()
        }
    })

    it('should handle protocol not found', async () => {
        const gcli = await gConnect()
        const cli = new poogClient(gcli, statusToError)
        try {
            await cli.poogUint()
            throw new Error('expected poogUint to throw')
        } catch (e) {
            expect(e).toBeInstanceOf(ErrProtocolNotFound)
            const es = e as ErrProtocolNotFound
            expect(es.protocol.value).toBe(poogProtocolID.value)
        } finally {
            gcli.close()
        }
    })

    afterAll(async () => {
        if (srv) {
            await srv.stop()
        }
    })
})
